scripts/analyze_codebase.py

```
@@ -0,0 +1,284 @@
+            
+            if code_hash in block_hashes:
+                if code_hash not in duplicates:
+                    duplicates[code_hash] = [block_hashes[code_hash]]
+                duplicates[code_hash].append(block_info)
+            else:
+                block_hashes[code_hash] = block_info
+    
+    return duplicates
+
+
+def find_unreferenced_modules(root_dir: Path, exclude_patterns: List[str]) -> Dict[str, Any]:
```

Ruff (0.12.2)

`find_unreferenced_modules` is too complex (14 > 10)

(C901)